doc/refactoring_design.md

```diff
@@ -0,0 +1,481 @@
+* In the end of current block interval, Ocean network randomly choose a winner from the providers who hold `lottery ticket`.
+* This chosen provider receives all tokens in the reward pool;
+* The procedure will repeat for each block interval; if there is no winner, Ocean tokens in the reward pool will accumulate over time until a winner claims the reward.
+
+<img src="img/refactoring/rewardLogic.jpg" width="1000" />
+
+The process within one block interval can be illustrated as below:
+
+<img src="img/refactoring/distributeBR.jpg" width="800" />
+
+
+
+The sample smart contract looks like this:
+
+```Solidity
+// providers of data commons in this block interval
+address[] candidates;
+
+// number of reward tokens in this block interval
+uint256 rewardPool;
+
+// randomly pick a provider to receive block reward
+function sendBlockReward() public returns (bool success){
+	...
+	// Ocean token contract mints new tokens
+	OceanTokenContract.mintTokens();
+	
+	// randomly pick a provider from the candidates (need a secure Random Number Generator!!!)
+	uint256 random = keccak256(abi.encodePacked(...)) % n;
+	
+	// send tokens in rewardPool to a randomly chosen provider
+	transfer(candidates[random], rewardPool);
+	...
+};
+```
+
+
+### 2.3 Service Agreement
+Please see [SLA.md](https://github.com/oceanprotocol/dev-ocean/blob/feature/SLA-specs/doc/sla.md) and [smart contract prototype](https://github.com/oceanprotocol/keeper-contracts/blob/develop/contracts/SLA/ServiceAgreement.sol) for details. 
+
+The structre of a Serivce Agreement looks like below:
+
+<img src="img/refactoring/SAStructure.jpg" width="700" />
+
+The key point is to split up the conditions from reward payment as:
+<img src="img/refactoring/SAOverview.jpg" width="1000" />
+
+
+**Some changes to be added:**
+
+
+1. **`Time Lock` and `Time Out`:** Each service condition has `time-lock` and `time-out`. It can only be fulfilled within the time window defined by the starting time point (`time-lock`) and the ending time point (`time-out`).
+<img src="img/refactoring/lifecycle.jpg" width="800" />
+
+2. **support Tri-State: Unknown, True, False**
+<img src="img/refactoring/triState.jpg" width="800" />
+
+3. **New service agreement template must go through governance and TCR to be whitelisted**. 
+<img src="img/refactoring/whitelistSAT.jpg" width="800" />
+
+4. **Add current provider of data commons into block reward candidates if service agreement is fulfilled and verified.**
+
+
+The sample smart contract looks like below:
+
+```Solidity
+// setup new whitelisted SA template and add it into Tribe and Curation contract
+function setupAgreementTemplate(address[] contracts, bytes4[] fingerprints, uint256[] dependenciesBits, bytes32 service) public returns (bool){
+	...
+	// service agreement template must be whitelisted at this time
+	require(service.isWhitelisted == true);
+	...
+	// add SA template into Tribe contract and Curation contract
+	OceanTribe.addServiceTemplate(bytes32 templateId);

```

Comment:
This makes the code tightly coupled. You just need to pass the SLA template ID to governance module and governance module will return T/F.